pulsar-client-tools/src/main/java/org/apache/pulsar/shell/PulsarShell.java

```diff
@@ -414,18 +441,34 @@ public List<String> readCommand() {
                 return;
             }
             execState = ExecState.RUNNING;
-            final String line = words.stream().collect(Collectors.joining(" "));
+            String line = words.stream().collect(Collectors.joining(" "));
             if (StringUtils.isBlank(line)) {
                 continue;
             }
             if (isQuitCommand(line)) {
+                if (shellMode != ShellMode.DEFAULT) {
+                    output(String.format("Exiting from %s shell mode", shellMode.command), terminal);
+                    prompt = createPrompt(promptMessage);
+                    shellMode = ShellMode.DEFAULT;
+                    continue;
+                }
                 exit(0);
                 return;
             }
+            if (shellMode != ShellMode.DEFAULT) {
+                words.add(0, shellMode.command);
+                line = shellMode.command + " " + line;
+            }
             if (isHelp(line)) {
                 shellCommander.usage(System.out);
                 continue;
             }
+            ShellMode newShellMode = ShellMode.valueOfCommand(line.toLowerCase(Locale.ROOT));
+            if (newShellMode != null) {
```

Comment:
checking for this makes only sense when `shellMode == ShellMode.DEFAULT`. Switching the shell mode should only be possible at the top level.

Reply:
Yes, I updated it. Now switching will only be possible at the top level.